Add schema markup to Authors to improve search results.

nk4456542 opened the pull request:
<!-- What issue does this PR close? -->
Closes #3231 

<!-- What does this PR achieve? [feature|hotfix|fix|refactor] -->
Add schema markup tags for the authors and improves the search result.

### Technical
<!-- What should be noted about the implementation? -->
As suggested in the comment of the issue, [here](https://github.com/internetarchive/openlibrary/issues/3231#issuecomment-602078002). I have made the changes, although I was confused as to what should be the `itemprop` in line `72` of my made changes, should it be `author` or `name` property?

The second para of the comment(except the `subjects` property) asks us to change this file:https://github.com/internetarchive/openlibrary/blob/1e31269adcc7118642e591e9dcfc04f400bdf658/openlibrary/macros/SearchResultsWork.html#L40
I don't know if I should add some more markup in this file too, so I did not do any changes. If I should do it in this same PR, do tell me, I would be happy to do it.

The `alternateName` property was to applied to each of the alternate name, so I made some changes in the code in the `page.alternate_names` section. As we were using `.join` method there, I just changed it to a `for` loop.

### Testing
<!-- Steps for reviewer to reproduce/verify what this PR does/fixes. -->
I ran the test locally i.e, `docker-compose exec make test`, the test did not produce any errors, also I tested for the author `Mark Twain` author page which is present in the local instance, seemed to work fine.

Also I tried to test the code snippets in the [structured data testing tool](https://search.google.com/structured-data/testing-tool/u/0/), but I did not receive anything relevant given that it is missing the context of the page.

### Screenshot
<!-- If this PR touches UI, please post evidence (screenshots) of it behaving correctly. -->
It does not touch any UI, but still I have included the HTML elements of the developer console from the `Mark twain` page
![Screenshot from 2020-10-25 16-08-19](https://user-images.githubusercontent.com/57295877/97104783-854d8b80-16dc-11eb-8a73-0a3dd6514e19.png)


### Stakeholders
<!-- @ tag stakeholders of this bug -->
@tabshaikh

hornc commented on openlibrary/templates/type/author/view.html:
```diff
@@ -207,11 +211,11 @@ <h3>$:_('Links <span class="gray small sansserif">(outside Open Library)')</span
+                    <li><a itemprop="sameAs" href="$page.wikipedia">Wikipedia</a></li>
                 $for link in page.links:
-                    <li><a href="$link.url">$link.title</a></li>
+                    <li><a itemprop="sameAs" href="$link.url">$link.title</a></li>
```
I don't think we can say that any link on an author bio here will be `sameAs` , it could be a personal blog or a storefront or somethine esle. VIAF and ISNI links above should probably be `sameAs` though.

nk4456542 replied:
will make these changes

hornc commented on openlibrary/templates/type/author/view.html:
```diff
@@ -194,7 +198,7 @@ <h6 class="collapse black uppercase">$label</h6>
             </div>
 
-        <div class="section">
+        <div class="section" itemscope itemtype="https://schema.org/Thing">
```
shouldn't this be `itemtype="https://schema.org/Person"` ?

nk4456542 replied:
Oh ok, will change it. But as mentioned in the comment [here](https://github.com/internetarchive/openlibrary/issues/3231#issuecomment-602078002) that all authors are not people, so I went for `Thing`. But will change it :).

nk4456542 commented:
@hornc, made the changes, give a look when you are free :)

hornc commented:
Thanks @nk4456542 for linking to the discussion. There is an `entity_type: org` that is present on many non-human authors, and should be on more. Example: https://openlibrary.org/authors/OL6821473A.json

(documented in the schema here: https://github.com/internetarchive/openlibrary-client/blob/83a37b2646c5ddda31e9ddc48d8401ed5fb50f4f/olclient/schemata/author.schema.json#L37-L40 )

That could be used (as well as presence of dates) to determine whether the entity is a Person or 'Thing'

Distinguishing between People and organisations is something I'll think about for the RDF improvements I've started in #3855

nk4456542 commented:
@hornc, I have added some code to check if the author is a `Person` or `Organization`. Schema markup was available for `Organization` so I have added https://schema/Organization. All these changes were based on [Walker House](https://openlibrary.org/authors/OL6821473A.json) and [Mark Twain](https://openlibrary.org/authors/OL18319A.json), Also I gave a look at some other authors' JSON. Seemed my code could work for that too. I don't know any other authors whose `entity_typ` is `org`. I have also changed the schema link for `alternateName`, I think it would go with the `entityTypeShema`. Have a look when you are free.

cdrini commented on openlibrary/templates/type/author/view.html:
```diff
@@ -68,15 +68,19 @@
-            $page.date
-    </h2>
+    <div itemscope itemtype="https://schema.org/Person">
+        <h1 itemprop="author">$title</h1>
```
Author isn't expected on type `Person`:

```suggestion
        <h1 itemprop="name">$title</h1>
```

nk4456542 replied:
I was confused about this, thanks for clearing it out. will make the required changes.

cdrini commented on openlibrary/templates/type/author/view.html:
```diff
@@ -175,10 +179,10 @@ <h2 class="collapse">
             $if subjects:
-                <div class="section">
+                <div class="section" itemscope itemtype="https://schema.org/CreativeWork">
                     <h6 class="collapse black uppercase">$label</h6>
                     $for subject, count in subjects:
-                        <a href="/subjects/$prefix$subject.lower().replace(' ', '_')">$subject</a>$cond(not loop.last, ",", "")
+                        <a itemprop="about" href="/subjects/$prefix$subject.lower().replace(' ', '_')">$subject</a>$cond(not loop.last, ",", "")
```
Having `CreativeWork` here doesn't seem correct; the closes thing looks like having `itemprop="knowsAbout"`. So let's remove the itemtype CreativeWork, and change the `itemprop="about"` to `knowsAbout` (See https://schema.org/Person )

cdrini commented on openlibrary/templates/type/author/view.html:
```diff
@@ -220,9 +229,12 @@ <h3>$:_('Links <span class="gray small sansserif">(outside Open Library)')</span
         $if page.alternate_names:
             $if page.alternate_names != [""]:
-                <div class="section">
+                <div class="section" itemscope itemtype=$entityTypeSchema>
```
This needs to get at the very top of the page, where you have `itemtype="https://schema.org/Person"`

```suggestion
                <div class="section">
```

cdrini reviewed (changes requested):
Looking good! a few changes to fix this up. Note I copy/pasted the page's HTML into https://search.google.com/structured-data/testing-tool/u/0/ to see what structured data it found.

We also want to add:

- `itemprop="image"` to the author's image
- `itemprop="description"` to the author's description

Note this is now on staging.openlibrary.org, so you should have more books to test :)  I tested using http://staging.openlibrary.org/authors/OL8624452A/John_Kaprielian

nk4456542 commented:
@cdrini, made the changes, give a look when you are free. The image is being rendered from the `macros/author_photo.html` page, but I was not sure if I should have added `image` itemproperty to the `authors/view` file or the `macro` file. But for now I have added in the `authors/view` page.

cdrini commented on openlibrary/templates/type/author/view.html:
```diff
@@ -64,20 +64,29 @@
-        $if page.date:
-            $page.date
-    </h2>
+    <div itemscope itemtype=$entityTypeSchema>
```
We need this to go higher up still, so that there is only one `Person` on this page, and all the `itemprop`s apply to it. It looks like we need this to go onto the `body`.

Something like this should do it:

```
$ bodyattrs = ctx.setdefault('bodyattrs', [])
$ bodyattrs.append('itemscope')
$ bodyattrs.append('itemtype="%s"' % entityTypeSchema)
```

Then in `body.html`, add:

```
$ bodyattrs = ctx.get('bodyattrs', [])
<body id="$bodyid" class="$(' '.join(bodyclass))" $(' '.join(bodyattrs))> 
```

(I.e. do the equivalent of what was done here: https://github.com/internetarchive/openlibrary/commit/6c5dabbe9bd4e54b2b1a7e629a25369b05c863e3 and here: https://github.com/internetarchive/openlibrary/commit/906541efe925f2159d7254a690b63d597fc0657f )

cdrini commented on openlibrary/templates/type/author/view.html:
```diff
@@ -64,20 +64,29 @@
 </div>
-<div id="contentBody">
+
+<div id="contentBody" itemscope itemtype=$entityTypeSchema>
```
This will define a second `Person`; once we move the `itemscope` to the `body`, this should no longer be needed :+1:

```suggestion
<div id="contentBody">
```

cdrini commented on openlibrary/templates/type/author/view.html:
```diff
@@ -168,7 +179,7 @@ <h2 class="collapse">
     </div>
     <div class="contentOnethird">
-        <div class="illustration">
+        <div class="illustration" itemprop="image">
```
You are correct, I believe this needs to go into `covers/author_photo`. It should go on a `<img>` element.

cdrini reviewed (changes requested):
Nice! 

1. Let's bring this down to just 1 `itemtype="...Person"`; otherwise the structured data sees that as 2 separate people!
2. Move `itemprop="image"` onto an `<img>` element

nk4456542 commented:
OMG, I am really sorry about the commits, I pulled from master and rebased it to the current branch after committing my changes :( 
That is the reason all the pulled commits are being shown. How do I undo this :( @cdrini , really sorry about this. Should I open a new PR and close this instead?

cdrini commented:
No worries! Git happens :) I was able to clear it up for you :+1: If you're curious, here's how:

```sh
git checkout master
git pull upstream master
git checkout -b nk
# Delete the last 50 requests; just gives us a clean canvas (avoid conflicts/merges)
git reset --hard HEAD~50
# Pull the commits from your branch
git pull https://github.com/nk4456542/openlibrary.git 3231/feature/add-schema-markup-to-authors-page

git rebase -i master
# Noticed there were a number of duplicated commits! Not sure how that happened, but I deleted the duplicates
# And the rebase went through without issue :)

# Confirm just the desired commits are here
git log master..HEAD

# Force push my local branch (nk) onto your fork's branch
git push -f https://github.com/nk4456542/openlibrary.git nk:3231/feature/add-schema-markup-to-authors-page
```

You can now update your local copy of the branch with

```sh
git fetch origin 3231/feature/add-schema-markup-to-authors-page
git checkout 3231/feature/add-schema-markup-to-authors-page
git reset --hard origin/3231/feature/add-schema-markup-to-authors-page
```

cdrini commented on openlibrary/templates/covers/author_photo.html:
```diff
@@ -15,12 +17,14 @@
     <div class="author coverMagic">
         <div class="SRPCover bookCover">
-            <img src="/images/icons/avatar_author-lg.png" class="cover" alt="We need a photo of $author.name"/>
+            <img src="/images/icons/avatar_author-lg.png" itemprop="image" class="cover" alt="We need a photo of $author.name"/>
```
In this case, since the image is just a placeholder image, we likely _don't_ want it to be itemprop=image
```suggestion
            <img src="/images/icons/avatar_author-lg.png" class="cover" alt="We need a photo of $author.name"/>
```

nk4456542 replied:
Will remove it :+1:

cdrini commented on openlibrary/templates/type/author/view.html:
```diff
@@ -38,6 +38,15 @@
+
+$ bodyattrs = ctx.setdefault('bodyattrs', [])
+$ bodyattrs.append('itemscope')
+$ bodyattrs.append('itemtype=%s' % entityTypeSchema)
```
Not _technically_ necessary, but good practice to have quotes around HTML attribute values.

```suggestion
$ bodyattrs.append('itemtype="%s"' % entityTypeSchema)
```

nk4456542 replied:
@cdrini, really thank you for cleaning that up, appreciate it a lot, you made my day today, to be honest, I will make sure it never happens again. :) 

Actually, I did that on purpose, because the [Structured Data Testing Tool](https://search.google.com/structured-data/testing-tool/u/0/) gives the error in the below picture, it is because double quotes are being rendered in the HTML body tag. I wanted to mention in the comment but I forgot to add after all that git stuff.

Should I add quotes or leave it as it is?

![Screenshot from 2020-11-18 03-30-35](https://user-images.githubusercontent.com/57295877/99456177-f6d0d080-294e-11eb-88ad-284c2c5cc508.png)
![Screenshot from 2020-11-18 03-31-10](https://user-images.githubusercontent.com/57295877/99456183-f9cbc100-294e-11eb-9361-28c0f4530692.png)

cdrini reviewed (approved):
Some small changes; lgtm! The structured data result is looking good :+1:

![Screenshot_2020-11-17 Structured Data Testing Tool](https://user-images.githubusercontent.com/6251786/99454091-eb9c8700-28f3-11eb-8aab-9f23f6c34cc7.png)

cdrini commented:
Applied those fixes; I'll merge once Travis is done testing :+1:
